zkevm-circuits/src/evm_circuit/util.rs

```diff
@@ -209,11 +209,14 @@ pub(crate) mod select {
 
 /// Decodes a field element from its byte representation
 pub(crate) mod from_bytes {
-    use crate::{evm_circuit::param::MAX_BYTES_FIELD, util::Expr};
+    use crate::{evm_circuit::param::MAX_N_BYTES_INTEGER, util::Expr};
     use halo2::{arithmetic::FieldExt, plonk::Expression};
 
     pub(crate) fn expr<F: FieldExt, E: Expr<F>>(bytes: &[E]) -> Expression<F> {
-        assert!(bytes.len() <= MAX_BYTES_FIELD, "number of bytes too large");
+        assert!(
+            bytes.len() <= MAX_N_BYTES_INTEGER,
+            "number of bytes too large"
+        );
```

Comment:
Yes there are, but we never know if it fits, until we check the whole number is less than the modulus. This is more like a warning to developer, if you are combining more than `31` bytes, you should know it's possible to wrap around. If we really need such feature, we can add function like `expr_unchecked` or mod like `from_bytes_unchecked`.

Reply:
If it's an assertion for the developer, shouldn't we make it correct for the entire ring using for example:
https://github.com/appliedzkp/pairing/blob/3f668ba2eab444ad9f4c6396db8292fada6feb42/src/bn256/fr.rs#L649

And failing if the `CtOption` is none. Which means the number is not canonical or over the modulus.

Reply:
But this is more like assertion for verifier, to warn it that with such usage, the value might wrap around, and we should never allow it.

Reply:
Hmmm. If we want to warn then probable we should use `debug_assert_eq!` maybe?

See: https://doc.rust-lang.org/std/macro.debug_assert.html

I say it because this is also failing for a small set of valid values.

Reply:
Just had a pairing session with @ed255 and we hit this assert on a case where the value was correct.

The workarround was to constrain the MSB to be 0 (as we were working with an address) and pass 31 bytes to the fn's with the assert.

Edu likes the assertion. So then we can keep it and close this discussion.

Apologies for the noise.

Reply:
> Just had a pairing session with @ed255 and we hit this assert on a case where the value was correct.

Ah, if you are saying about #287, in such case, the address is already known 20 bytes to verifier, so we only need 20 bytes for decomposition, and then use random linear combination on it. Then we don't need to query extra 12 cells and specially constrain them to be zero.

> If we want to warn then probable we should use debug_assert_eq! maybe?

There are many raw `assert!` is used in the circuit, and I think using `debug_assert!` makes much more sense. I have filed #288 and will try to resolve in future PRs.